tsdb/shard.go

```diff
@@ -702,56 +702,64 @@ func (s *Shard) validateSeriesAndFields(points []models.Point, tracker StatsTrac
 			continue
 		}
 
-		name := p.Name()
-		mf := engine.MeasurementFields(name)
-
-		// Check with the field validator.
-		if err := ValidateFields(mf, p, s.options.Config.SkipFieldSizeValidation); err != nil {
-			switch err := err.(type) {
-			case PartialWriteError:
-				if reason == "" {
-					reason = err.Reason
+		cont, err := func(p models.Point, iter models.FieldIterator) (cont bool, err error) {
+			name := p.Name()
+			mf := engine.MeasurementFields(name)
+			mf.mu.Lock()
```

Comment:
It feels like there is still a potential race condition. We lock the `mf` while looking up fields here, but then unlock it while we continue to the next point. Another incoming write in a different goroutine could then look up fields in `mf` before this goroutine can create the new fields. Or am I missing something?

Reply:
The field creation has its own locking and checks yet again for field type conflicts. So either of the go routines may win, and the other will report an error.

So, the race you describe is real, but gets sequenced in field creation.

Reply:
I changed the `MeasurementFields` lock to a `RWMutex` and used an `RLock` in  `Shard.validateSeriesAndFields` to allow greater parallelism. 

The lock has to be taken on a per point basis, because points can differ in which measurement they pertain to, and thus which `MeasurementFields` object has to be locked.

 `MeasurementFields.CreateFieldIfNotExists` still uses the full lock, of course.